### PR TITLE
ignore router get return 404  error

### LINKED ISSUE #557: do not fail deleting router if it does not exist
/kind bug

**What steps did you take and what happened:**
 - deleted router from openstack
 - deleted cluster

**What did you expect to happen:**
 - CAPO to realize that router does not exist and continue cleaning up remaining resources

**Anything else you would like to add:**
```
I0610 12:36:28.521243       1 openstackcluster_controller.go:113] controllers/OpenStackCluster "msg"="Reconciling Cluster delete" "cluster"="capi-quickstart" "namespace"="default" "openStackCluster"="capi-quickstart" 
I0610 12:36:29.126094       1 openstackcluster_controller.go:156] controllers/OpenStackCluster "msg"="Deleting router" "cluster"="capi-quickstart" "namespace"="default" "openStackCluster"="capi-quickstart" "name"="k8s-clusterapi-cluster-default-capi-quickstart"
E0610 12:36:29.542064       1 controller.go:258] controller-runtime/controller "msg"="Reconciler error" "error"="failed to delete router: Resource not found"  "controller"="openstackcluster" "request"={"Namespace":"default","Name":"capi-quickstart"}
I0610 12:39:13.382433       1 openstackcluster_controller.go:113] controllers/OpenStackCluster "msg"="Reconciling Cluster delete" "cluster"="capi-quickstart" "namespace"="default" "openStackCluster"="capi-quickstart" 
I0610 12:39:13.997013       1 openstackcluster_controller.go:156] controllers/OpenStackCluster "msg"="Deleting router" "cluster"="capi-quickstart" "namespace"="default" "openStackCluster"="capi-quickstart" "name"="k8s-clusterapi-cluster-default-capi-quickstart"
E0610 12:39:14.526687       1 controller.go:258] controller-runtime/controller "msg"="Reconciler error" "error"="failed to delete router: Resource not found"  "controller"="openstackcluster" "request"={"Namespace":"default","Name":"capi-quickstart"}
...
```

**Environment:**

- Cluster-api version: v0.3.0
- Minikube/KIND version
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jichenjc**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #557 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by hidekazuna** [comment]
@jichenjc looks good to me.. Do we have to fix other deleting resources?

**Comment by jichenjc** [comment]
not sure, I will check others but I can't test all of them :)

so I will see what I can do to improve and if further bugs reported, we can fix it as well

File: pkg/cloud/services/networking/securitygroups.go [hunk 1]
```diff
@@ -295,6 +297,9 @@ func (s *Service) exists(groupID string) (bool, error) {
 	}
 	allGroups, err := groups.ExtractGroups(allPages)
 	if err != nil {
+		if capierrors.IsNotFound(err) {
```
Comment (sbueringer) [inline pkg/cloud/services/networking/securitygroups.go:300]: Does ExtractGroups really throw these errors? I would have guessed they are returned by "network" calls like Get/List etc?

File: pkg/utils/errors/errors.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	if errCode, ok := err.(gophercloud.ErrUnexpectedResponseCode); ok {
+		if errCode.Actual == http.StatusNotFound {
+			return true
+		}
+	}
+
+	return false
+}
+
+func IsInvalidError(err error) bool {
```
Comment (sbueringer) [inline pkg/utils/errors/errors.go:43]: unused right now?

Comment (jichenjc) [reply pkg/utils/errors/errors.go:43]: yes, just in case will be used later :)

File: pkg/cloud/services/networking/securitygroups.go (hunk 1 above)
Comment (jichenjc) [reply pkg/cloud/services/networking/securitygroups.go:300]: um... I didn't test this, so I will remove this check 
after further test , we can add through another PR

File: pkg/utils/errors/errors.go (hunk 2 above)
Comment (sbueringer) [reply pkg/utils/errors/errors.go:43]: kk

**Pull request merged**